src/DataService/CachingPhpDocExtractor.php

```diff
@@ -0,0 +1,245 @@
+
+        switch ($source) {
+            case self::PROPERTY:
+                $tag = 'var';
+                break;
+
+            case self::ACCESSOR:
+                $tag = 'return';
+                break;
+
+            case self::MUTATOR:
+                $tag = 'param';
+                break;
+        }
+
+        $types = [];
+        /** @var DocBlock\Tags\Var_|DocBlock\Tags\Return_|DocBlock\Tags\Param $tag */
+        foreach ($docBlock->getTagsByName($tag) as $tag) {
+            if ($tag && null !== $tag->getType()) {
+                $types = array_merge($types, $this->phpDocTypeHelper->getTypes($tag->getType()));
+            }
+        }
+
+        if (!isset($types[0])) {
+            return;
+        }
+
+        if (!in_array($prefix, $this->arrayMutatorPrefixes)) {
+            return $types;
+        }
+
+        return [new Type(Type::BUILTIN_TYPE_ARRAY, false, null, true, new Type(Type::BUILTIN_TYPE_INT), $types[0])];
+    }
+
+    /**
+     * Gets the DocBlock for this property.
+     *
+     * @param string $class
+     * @param string $property
+     *
+     * @return array
+     */
+    private function getDocBlock($class, $property)
+    {
+        $propertyHash = sprintf('%s::%s', $class, $property);
+
+        if (isset($this->docBlocks[$propertyHash])) {
+            return $this->docBlocks[$propertyHash];
+        }
+
+        $ucFirstProperty = ucfirst($property);
+
+        try {
+            switch (true) {
+                case $docBlock = $this->getDocBlockFromProperty($class, $property):
+                    $data = [$docBlock, self::PROPERTY, null];
+                    break;
+
+                case list($docBlock) = $this->getDocBlockFromMethod($class, $ucFirstProperty, self::ACCESSOR):
+                    $data = [$docBlock, self::ACCESSOR, null];
+                    break;
+
+                case list($docBlock, $prefix) = $this->getDocBlockFromMethod($class, $ucFirstProperty, self::MUTATOR):
+                    $data = [$docBlock, self::MUTATOR, $prefix];
+                    break;
+
+                default:
+                    $data = [null, null, null];
+            }
+        } catch (\InvalidArgumentException $e) {
+            $data = [null, null, null];
+        }
+
+        return $this->docBlocks[$propertyHash] = $data;
+    }
+
+    /**
+     * Gets the DocBlock from a property.
+     *
+     * @param string $class
+     * @param string $property
+     *
+     * @return DocBlock|null
+     */
+    private function getDocBlockFromProperty($class, $property)
+    {
+        // Use a ReflectionProperty instead of $class to get the parent class if applicable
+        try {
+            $reflectionProperty = new \ReflectionProperty($class, $property);
+        } catch (\ReflectionException $e) {
+            return;
+        }
+
+        return $this->docBlockFactory->create($reflectionProperty, $this->contextFactory->createFromReflector($reflectionProperty->getDeclaringClass()));
```

Comment:
And, the one use of `$this->contextFactory` is _also_ in a private method, so once we would have to override all of the public methods we may as well just copy the whole class